### ISSUE TITLE
Leveraging pdk, garbage collector builds up under certain plugin/phase execution on Kong 1.0/014.1

### ISSUE THREAD

**Issue by jeremyjpj0916** [Kong/kong#4143]
Edited, see later posts.

**Comment by bungle** [comment]
@jeremyjpj0916 Thanks for the report. Surely something we start to look asap. Another thing to try is to run without our patches[1]. Our binaries come patched, so you need to run e.g. vanilla OpenResty `1.13.6.2`. Thank you for reporting.

[1] https://github.com/kong/openresty-patches

**Comment by jeremyjpj0916** [comment]
@bungle No problem, I don't want to have yall go on a wild goose chase for nothing though so I am investigating it further and will keep this issue updated with any findings. Following the life of the POD (it has taken no active traffic now besides healthchecks pings since my first post) we can see it has grown a bit further but it seems finally in the last couple hours it may have plateaued and reached a sort of stasis:

![image](https://user-images.githubusercontent.com/31913027/50576798-80d22780-0e11-11e9-8284-2eebaaac9242.png)
 
How Kong went from 435 MiB to about 568 MiB peak I am not sure exactly(just regular alpine linux runtime growth/openresty/nginx memory needs? I top'ed the worker processes again with the above and see them now at the amount pictured below, so each grew a good chunk during this idle period as well), 130 MiB growth over time but if it does not continue its upward climb then I would begin to rule out core having a nasty memory leak.

![image](https://user-images.githubusercontent.com/31913027/50576849-bcb9bc80-0e12-11e9-8205-2505dff93fb8.png)


As for my Prod 0.14.1 environment with heavy leak the plugins I will be most suspicious of currently would probably be:
1. https://github.com/Optum/kong-error-log 
2. https://github.com/Optum/kong-splunk-log (Based mostly on Kong's newer HTTP Log Plugin logic).

**Comment by jeremyjpj0916** [comment]
@p0pr0ck5 will do some tracking on that from a fresh deploy starting tomorrow probably. I want to continue to let this guy run and see if it indeed has reached a flatline. As for current VmRSS(current resident set) here you are on a few of them:

![image](https://user-images.githubusercontent.com/31913027/50577440-edecb980-0e1f-11e9-9111-f82262658ecb.png)

**Comment by jeremyjpj0916** [comment]
@bungle @p0pr0ck5 aight friends mystery is solved. 

Good news: Seems Kong 1.0 barebones seems fine. After it reached that plateau after many hours I saw no more arbitrary climbing in mem for no reason. I am editing my title to be more appropriate, no need to wrongly rain on the 1.0 new release parade for no reason at this point 😆 .

Bad news: Kong 0.14.1 is impacted, and it was indeed actually a very specific use case that can reproduce it(1.0 is actually safe from this exact situation currently it seems from initial tests due to refactors), a use case I still personally consider fairly common in the real world. Finding the problem in 0.14.1 and understanding where the leak occurred may save Kong in the future for implementation that I have not yet tested through, even in 1.0 possibly. Lets review what causes it and the code path it must take to produce the mem leak.

1. Run 0.14.1 , create a consumer and give him oauth creds , create a service+route, add the acl plugin with whitelist and add the oauth2 plugin with client credentials + global options enabled. Make sure the consumers acl gets the whitelist so he can call the proxy. Generate a token so the user can call the proxy.

2. The exact situation is indeed as I was describing earlier here:

> Edit - I actually have a crazy strange hunch that recently we had XML driven API calls start occurring more frequently in prod for a few teams heavier load testing and right around then is when I began to notice decent growth in our pod RAM consumption. Ofc these tx's which produce this err in the logs 1,000's of times per day constantly on 0.14.1.

2019/01/02 02:35:28 [error] 36#0: *6728866 [lua] public.lua:139: get_body_info(): don't know how to parse request body (can't decode Content-Type 'text/xml'), client: 10.***.*.***, server: kong, request: "POST /api/some/path/service-view/v1/name HTTP/1.1", host: "gateway.company.com"

Let me see if I can get public.lua screaming at me again on 1.0, I know on 0.14.1 a proxy protected with acl/oauth2 client credentials with an XML call produced the above due to lame attempted reading of the body to get the token even though my oauth2 bearer token was in the headers(which I submitted a PR for so 1.0 will look at headers first thankfully without req body reading). But if I start putting the bearer token in the body and oauth2 then is forced to read from the body(which I will send in XML), lets see if this error path within Kong is whats causing me the leak


Now to produce that error yourself just send a POST call like this example, do application/xml as your content type.
```
POST https://gateway-core-dc2.company.com/jeremy/memleak?access_token=75NEJ2k5j1g3OJeFo4SqE74sDtJofmlq

<soap:Envelope
xmlns:soap="http://www.w3.org/2003/05/soap-envelope/"
soap:encodingStyle="http://www.w3.org/2003/05/soap-encoding">
<soap:Body>
  <m:GetPrice xmlns:m="https://www.w3schools.com/prices">
    <m:Item>Apples</m:Item>
  </m:GetPrice>
</soap:Body>
</soap:Envelope>
```

Now the code path that kicks off the mem leak on 0.14.1 can be found here:
https://github.com/Kong/kong/blob/bfbb0d3822b31f0d5da401acbb19d29d10251b8a/kong/plugins/oauth2/access.lua#L111

And then starts into the public.lua file:
https://github.com/Kong/kong/blob/bfbb0d3822b31f0d5da401acbb19d29d10251b8a/kong/tools/public.lua#L151

So something around when this error gets constantly thrown in public.lua is when the memory leak begins:
```
2019/01/02 05:46:32 [error] 216#0: *4390028 [lua] public.lua:139: get_body_info(): don't know how to parse request body (can't decode Content-Type 'application/xml'), client: 10.***.***.***, server: kong, request: "POST /jeremy/memleak?access_token=75NEJ2k5j1g3OJeFo4SqE74sDtJofmlq HTTP/1.1", host: "gateway-stage-core-dc2.company.com"
```

First sort of attack vector I have uncovered too in a way. Because if you were to discover someones Kong OAuth proxy endpoint url, and send a bad token with xml payload rapidly, that public_utils.get_body_args() from the OAuth logic will still get invoked and the memory will skyrocket.

My test was ran at 120 TPS just now doing this and you can see how it impacted mem growth, as soon as the below line started curving up was when I ran my test:
![image](https://user-images.githubusercontent.com/31913027/50583264-13011c80-0e60-11e9-97e2-daf2e8e9ea4f.png)

Finally, why does this not impact Kong 1.0. Well Kong 1.0 does not use the public_utils lib call any longer and is not logging the above error I have been mentioning to stderr on proxy calls:
https://github.com/Kong/kong/blob/master/kong/plugins/oauth2/access.lua#L130 

Plus it has my PR at this point, so if you are using OAuth with the token in a header the body call read's never occur in the first place.

Welp been decent fun with the investigative work. Now I am too tired to want to learn stapxx haha. Maybe yall will see in public.lua the bad code that does grow the mem. If you do let me know! Should be able to reproduce this pretty easily with the ammo I have just given.

**Comment by hishamhm** [comment]
@jeremyjpj0916 that `get_body_args` causes the body to be read in its entirety into Lua-land and interned as a string. I don't see it hoarding that data in any structure so at some point it should be GC'd... so maybe it's just that the garbage collection rate is lower than the garbage production rate, which would cause the indefinite memory growth.

**Comment by jeremyjpj0916** [comment]
Well @hishamhm you may have a point, I still need to keep researching but just testing an abritrary proxy with no auth or anything seems to cause growth as well at 102 TPS, an example of my stage environment showing the growth, you can tell where the test started:

![image](https://user-images.githubusercontent.com/31913027/50721200-65864700-10b3-11e9-8a3d-f9edcb097f2c.png)

I am adding a snippit to my custom kong template too in admin api section:
```
      location /gc {
        content_by_lua_block {
            ngx.say(string.format("Worker %d: GC size: %.3f KB", ngx.var.pid, collectgarbage("count")))
        }
      }
```
So I can see what lua gc is reporting the size to be to see if its all that. I also figured out how to install and run kong with valgrind memory debug on, maybe I can pipe that output to a file and then search for the memory leak keywords they explain I would see if it detected a mem leak. 

I could not figure out a way to install stapxx on alpine linux, there are a few libs it seems not supported on alpine vs debian. I suppose if I get desperate enough I will convert to debian deploys temporarily in dev if stapxx would give better insight.

My next real question would be is if it is the fact gc can't keep pace @ 100 tps with our plugins, is there absolutely anything we can do about it to enforce a quicker cleanup of mem? I would think as quick as a tx goes through and completes any buffers and lua references that are only needed for that specific call would be freed back up for another. I can't disable reading the request body(although in dev I may do just that to see if it actually even helps if I can reproduce the leak in a dev).

Gonna continue to try things and see what happens, currently I am just redeploying prod every 3 days as mem fills up to prevent the "Terminated at Dec 30, 2018 11:21:57 AM with exit code 137 (OOMKilled)" container crashes. 

Another thing I am pondering if I run 6 worker processes, each luaVM can use max 2GB then the minimum RAM I should be running should be around like 13-14 GB right? Maybe thats my problem we only a lot 10GB to the container and serious gc and cleanup does not occur till the luaVM "fills up", I am a novice when it comes to these things obviously 😆  .

**Comment by jeremyjpj0916** [comment]
Doing a 102 TPS loadtest on 1.0 with this payload to a no plugins proxy (besides my global ones, one of which reads the request body) yields this so far in growth with my gc checker method now. I am going to remove all payload reading and test as well too tomorrow:

![gcmemoryfootprintgains](https://user-images.githubusercontent.com/31913027/50722101-c61c8080-10c1-11e9-8cac-74da2b92c408.png)

**Comment by jeremyjpj0916** [comment]
Interesting findings folks, so I have been disabling globally enabled plugins on my dev environment on 1.0 where I was able to reproduce the mem growth. I was getting really close to chalking it up to gc not being able to keep up(but we are talking about like 100-200 TPS load-tests here), which I highly doubt is enough to cause grief like I was seeing). I had planned on modifying the gc to be more aggressive following my readings in the init worker phase(Still may come in handy sometime under extreme load potentially):
```
collectgarbage("setstepmul", 1000) --200 is the default, The step multiplier controls the relative speed of the collector relative to memory allocation. Larger values make the collector more aggressive but also increase the size of each incremental step. Values smaller than 100 make the collector too slow and can result in the collector never finishing a cycle. The default, 200, means that the collector runs at "twice" the speed of memory allocation.
collectgarbage("setpause", 110)		 --200 is the default, The garbage-collector pause controls how long the collector waits before starting a new cycle. Larger values make the collector less aggressive. Values smaller than 100 mean the collector will not wait to start a new cycle. A value of 200 means that the collector waits for the total memory in use to double before starting a new cycle.
```

But then it happened, one of my global plugins I had enabled I disabled and the results look really positive:
https://docs.konghq.com/hub/optum/kong-response-size-limiting/
https://github.com/Optum/kong-response-size-limiting/blob/master/src/handler.lua 

GC size never grows beyond 50,000kb at current load (network traffic inbound around 800KiB's per second) and stasis seems achieved:
![image](https://user-images.githubusercontent.com/31913027/50730660-82218e00-114a-11e9-8d5d-a4b42677683c.png)

Now the head scratcher is what within the plugin is causing the gc to go whack and allow such growth??? Based on all my readings I see nothing wrong in the plugin, it follows the pattern the request size limiter took that kong wrote and it follows setting the response based on some of the openresty documentation I saw. Could something sinister more underlying be causing gc to be rekt? I may try commenting out the body_filter vs the header_filter phases and isolating to see if its a specific phase or what causing it...

I can almost taste the finish line with this issue now and take a breath hah.

**Comment by jeremyjpj0916** [comment]
100% confirmed that its this plugin someway somehow. Even after I refactored it a bit cleaner to fully leverage the pdk gc goes through the roof under load with this plugin enabled globally:

https://docs.konghq.com/hub/optum/kong-response-size-limiting/
https://github.com/Optum/kong-response-size-limiting/blob/master/src/handler.lua

Editing my title accordingly to this github issue, something about this code and its phase execution causes gc to just stop doing its job well somehow and gc just keeps a growin. Editing my book of older posts to shrink down this github issue.

**Comment by p0pr0ck5** [comment]
Im suspicious of kong.ctx hereI’m curious, why this plugin needs to use kong.ctx? Any reason not to just exit in the header filter? If you factor out the usage of kong.ctx to just use ngx.ctx, or just avoid the body filter entirely, do you still see memory increases?

**Comment by p0pr0ck5** [comment]
This is an oddity within the PDK then, as the `exit` ngx Lua function can most assuredly be called from within the header filter phase: https://github.com/openresty/lua-nginx-module/blob/master/README.markdown#ngxexit.

**Comment by p0pr0ck5** [comment]
The error you’re seeing is not triggered by the lua nginx module, but by the PDK.

**Comment by jeremyjpj0916** [comment]
Yep so just setting ngx.exit with a status of 413 seems to work. Have not watched mem footprint yet but it looks like this on a response (which I don't like because its really the backend response entity I want to talk about not the request, and openresty does not give a way to do so until the body_filter phase:

```
HTTP/1.1 413
Server: openresty/1.13.6.2
Date: Mon, 07 Jan 2019 02:35:52 GMT
Content-Type: text/plain; charset=UTF-8
Transfer-Encoding: chunked
X-RateLimit-Remaining-second: 499
X-RateLimit-Limit-second: 500
optum-cid-ext: 0ae6dba6-221a-4686-86f6-c3f3e37fb8f6#1
X-Upstream-Latency: 6
X-Proxy-Latency: 34
Set-Cookie: 9b2731b6f344dac3aa25878b9b1193ea=4edba7b4836fe28e48c1124d52375bc7; path=/; HttpOnly; Secure

Payload too large
```

**Comment by jeremyjpj0916** [comment]
Thanks for the quick response, I will see if using ngx.ctx fixes this for me and go with my original logic(In the meanwhile I suppose yall will want to try reproducing the problem locally and figure out what/why causes this with the pdk). 

As for the server header, yes my config does currently not specify setting server headers:
![image](https://user-images.githubusercontent.com/31913027/50746417-dac65900-1226-11e9-8ae3-d0d6ca8b780b.png), ohhh or are you saying I can only overwrite openresty for kong header and not remove completely 😆 , guess I will have to get in there and hack it out. Not that I care for security through obscurity, just one less pointless header in my mind that doesn't need to be sent necessarily. But I understand a need for "branding" I suppose lol.

**Comment by jeremyjpj0916** [comment]
Preliminary analysis looks good. What is interesting is when producing the mem growth on the earlier code, I was calling the proxy with a payload size LESS than the size that would hit the code path that would set the kong.ctx.plugin.limited field to true in the header_filter phase. So basically the thing that I suppose is producing the mem growth is the if kong.ctx.plugin.limited check in my body filter on a var that would be nil????

**Comment by jeremyjpj0916** [comment]
Update, so I went with a purist approach and I think this seems to not reveal any leaks and to have supported a plateau of sorts under load. Let me know if you see anything wrong with the implementation:

```
local BasePlugin = require "kong.plugins.base_plugin"
local tonumber = tonumber
local MB = 2^20

local responsestr = "{\"message\": \"Response size limit exceeded\"}"
local sizetrigger = false

local KongResponseSizeLimitingHandler = BasePlugin:extend()
KongResponseSizeLimitingHandler.PRIORITY = 802

function KongResponseSizeLimitingHandler:new()
  KongResponseSizeLimitingHandler.super.new(self, "kong-response-size-limiting")
end

local function check_size(length, allowed_size)
  local allowed_bytes_size = allowed_size * MB
  if length > allowed_bytes_size then
      sizetrigger = true
      --kong.response.set_status(413)
      --kong.response.set_header("Content-Length", #responsestr)
      --kong.response.set_header("Content-Type", "application/json")
      ngx.status = 413
      ngx.header["Content-Length"] = #responsestr
      ngx.header["Content-Type"] = "application/json"
  end
end

function KongResponseSizeLimitingHandler:header_filter(conf)
  KongResponseSizeLimitingHandler.super.header_filter(self)
  local cl = ngx.header.content_length
  if cl and tonumber(cl) then
    check_size(tonumber(cl), conf.allowed_payload_size)
  else
    ngx.log(ngx.DEBUG, "Upstream response lacks Content-Length header!")
  end
end

function KongResponseSizeLimitingHandler:body_filter(conf)
  KongResponseSizeLimitingHandler.super.body_filter(self)
  if sizetrigger then		
    ngx.arg[1] = responsestr
    ngx.arg[2] = true
  end
end

return KongResponseSizeLimitingHandler
```

I think this github reveals there may be situations where the pdk needs revision around which phases support which methods, as well as potential memory growth/leak(not sure which) that cause the gc to grow quickly without clearing itself. @p0pr0ck5 if you feel you have enough information based on the revision history of my plugin in github and such for ammo feel free to close this issue whenever, with my latest plugin code at least for now not leveraging the pdk my plugin and Kong nodes can show stability(without scrapping this plugin all together or marking it off as some sort of openresty leak/isssue).

**Comment by p0pr0ck5** [comment]
I'd like to leave this open for now, at least until @bungle or @thibaultcha who have worked on the PDK have a chance to review. It's interesting to hear about the memory growth when hitting code paths wherein a `kong.ctx.plugin` key is nil; I wonder if there's some odd behavior in the `__index` definition?

**Comment by jeremyjpj0916** [comment]
Sounds like a plan. I can confirm after load testing both paths this plugin can take with my latest code in the repo and commented above, no mem leaks occur so at least that is good as a baseline. Was worried OR has issues with ngx.ctx between the header_filter to body_filter references around the garbage collecting hah. agentzh dodges a bullet with my Github rambles 😄 . Also can't believe I tried that local boolean late at night, I am dumb.

**Comment by thibaultcha** [comment]
I may have found a potential issue with `kong.ctx`, but I am having trouble reproducing your observations @jeremyjpj0916. Could you help us narrowing this down to either `kong.ctx` or `kong.response.get_header`? 
Also, when you were observing this, how many Routes/Services/other entities were configured on the node at the same time? How were you load testing Kong? From a single Route, or several at once? And how large is your `mem_cache_size` setting?

**Comment by jeremyjpj0916** [comment]
@thibaultcha Sup man, sure let me overwrite my sandbox real quick with the old codebase, I will comment out the body_filter and see if just the get_header produces it(I won't have the size be enough to trigger going in the if  statement even in the header_filter phase). If not then I would have to  say to say the kong.ctx.plugin.limited was what was doing the deed. To talk about where I saw the issue, on dev(sandbox)/stage/prod all environments had the mem growth occurring, and in stage/prod at the moment I just have the plugin disabled and now Kong runs at a steady 700MiBs of ram daily as opposed to eating through 10GB of RAM in 5-6 days of prod load traffic(stage has about 1k routes/service pairs, prod has about 50-100), sandbox i too could reproduce I had 4-5 routes/services when testing it. Seems that number of resources has no correlation to the problem imo.  As for testing method just Kong 1.0, with the plugin enabled globally and a bare route/service nothing on it(was actually routing to a real host that just responded with a Status success and empty body. mem_cache_size also did not seem to correlate, I usually had it set to 1024m in my environments but for sandbox I temporarily made that 1m too during testing to isolate out the shm growing potentially. I was loadtesting using SOAPUI locally on my work laptop at about 150 TPS, but then to crank up the load I also had wrk too on one of our linux hosts start sending traffic at another 150 TPS, the mem growth definitely can be associated with higher throughput(i don't think req or resp bodies make the real difference here honestly), which I think means just hitting the plugin path at a higher frequency shows the problem faster. But I would include a post body just because that is how I have been testing(as seen in posts from earlier).

**Comment by thibaultcha** [comment]
@jeremyjpj0916 I think I have found the root cause, no rush on testing this all over again on your side for now. Continuing my investigation for now, just wanted to avoid you some extra work.

**Comment by jeremyjpj0916** [comment]
I was already running the load test before you commented hah, didn't take but a second to setup and for all the diligent fixes yall provide its the least I can do to to retest and confirm a few extras. And yeah it seems growth still present and points to kong pdk getting the header on content-length when content-length isn't even present in the response headers. My gc size per worker is already over the 50,000kb size where I see it usually cap at when no mem growth occurs. So I think that is part one. And maybe after that fix the kong.ctx.plugin limited is not the issue but best to fix one part and retry.

**Comment by thibaultcha** [comment]
Yes, that corroborates the leak I have just found, which is separate from the `kong.ctx` issue I found earlier. Thanks for testing this quickly :)

**Comment by jeremyjpj0916** [comment]
Anytime, great to hear you found the problem. I will go back to piddling with other 1.0 things now. Will give my plugin a pdk refactor again once the pdk logic around what I have tried to do here has been confirmed fixed. Take care.

**Comment by thibaultcha** [comment]
PR: https://github.com/Kong/kong/pull/4172

@jeremyjpj0916 Mind giving it a spin? With `kong.ctx` _and_ `kong.service.response` of course :) Thanks!

**Comment by jeremyjpj0916** [comment]
@thibaultcha  Am doing,

handler.lua
```
local BasePlugin = require "kong.plugins.base_plugin"
local tonumber = tonumber
local MB = 2^20

local responsestr = "{\"message\": \"Response size limit exceeded\"}"

local KongResponseSizeLimitingHandler = BasePlugin:extend()
KongResponseSizeLimitingHandler.PRIORITY = 802

function KongResponseSizeLimitingHandler:new()
  KongResponseSizeLimitingHandler.super.new(self, "kong-response-size-limiting")
end

local function check_size(length, allowed_size)
  local allowed_bytes_size = allowed_size * MB
  if length > allowed_bytes_size then
      kong.ctx.plugin.responsetoolarge = true
      kong.response.set_status(413)
      kong.response.set_header("Content-Length", #responsestr)
      kong.response.set_header("Content-Type", "application/json")		
  end
end

function KongResponseSizeLimitingHandler:header_filter(conf)
  KongResponseSizeLimitingHandler.super.header_filter(self)
  local cl = kong.service.response.get_header("content-length")
  if cl and tonumber(cl) then
    check_size(tonumber(cl), conf.allowed_payload_size)
  else
    ngx.log(ngx.DEBUG, "Upstream response lacks Content-Length header!")
  end
end

function KongResponseSizeLimitingHandler:body_filter(conf)
  KongResponseSizeLimitingHandler.super.body_filter(self)
  if kong.ctx.plugin.responsetoolarge then
    ngx.arg[1] = responsestr
    ngx.arg[2] = true
  end
end

return KongResponseSizeLimitingHandler
```

My modified build patch work:
```
RUN git clone https://github.com/Optum/kong-response-size-limiting.git /usr/local/share/lua/5.1/kong/plugins/kong-response-size-limiting &&\
    find /usr/local/share/lua/5.1/kong/plugins/kong-response-size-limiting -type f -maxdepth 1 -delete &&\
    cd /usr/local/share/lua/5.1/kong/plugins/kong-response-size-limiting &&\
    mv src/* . &&\
    rm -r src

COPY patches/memleak/handler.lua /usr/local/share/lua/5.1/kong/plugins/kong-response-size-limiting/handler.lua
RUN chmod 777 /usr/local/share/lua/5.1/kong/plugins/kong-response-size-limiting/handler.lua

##############################################################
##### BEGIN Tibo memleak fix  #####

COPY patches/tibo/memleak/ctx.lua /usr/local/share/lua/5.1/kong/pdk/ctx.lua
RUN chmod 777 /usr/local/share/lua/5.1/kong/pdk/ctx.lua

COPY patches/tibo/memleak/response.lua /usr/local/share/lua/5.1/kong/pdk/service/response.lua
RUN chmod 777 /usr/local/share/lua/5.1/kong/pdk/service/response.lua

##### END Tibo memleak fix  #####
##############################################################
```


currently watching to see if it flat lines like when I use openresty ngx calls directly, still going up so far slowly:
![image](https://user-images.githubusercontent.com/31913027/50876417-d637a700-13c4-11e9-8df4-afcf8117cebb.png)

**Comment by jeremyjpj0916** [comment]
Follow up:

![image](https://user-images.githubusercontent.com/31913027/50877393-8c04f480-13c9-11e9-9db2-8517b42bd5d3.png)


I think it has reached stasis. Also GC seems to be at the right levels. I tested both code paths this plugin could take to hit the paths of the content-length grab and the kong.ctx* bool path. Good stuff. Can close this issue if you feel there has been adequate resolution and the PR can move forward to the next release.

**Comment by thibaultcha** [comment]
Closed following up the merge of #4172. Thanks for the report @jeremyjpj0916!
